demos/python_demos/object_detection_demo_ssd_async/object_detection_demo_ssd_async.py

```diff
@@ -15,239 +15,593 @@
+                         mode='constant', constant_values=0)
+        img = img.transpose((2, 0, 1))  # Change data layout from HWC to CHW
+        img = img.reshape((self.n, self.c, self.h, self.w))
+        inputs[self.image_blob_name] = img
+        return inputs, meta
+
+    def postprocess(self, outputs, meta):
+        detections = self.output_parser(outputs)
+        orginal_image_shape = meta['original_shape']
+        resized_image_shape = meta['resized_shape']
+        scale_x = self.w / resized_image_shape[1] * orginal_image_shape[1]
+        scale_y = self.h / resized_image_shape[0] * orginal_image_shape[0]
+        for detection in detections:
+            detection.xmin *= scale_x
+            detection.xmax *= scale_x
+            detection.ymin *= scale_y
+            detection.ymax *= scale_y
+        return detections
+
+
+class SingleOutputParser:
+    def __init__(self, all_outputs):
+        if len(all_outputs) != 1:
+            raise ValueError('Network must have only one output.')
+        self.output_name, output_data = next(iter(all_outputs.items()))
+        last_dim = np.shape(output_data)[-1]
+        if last_dim != 7:
+            raise ValueError('The last dimension of the output blob must be equal to 7, '
+                             'got {} instead.'.format(last_dim))
 
     def __call__(self, outputs):
-        return outputs[self.output_layer].buffer[0][0]
+        return [Detector.Detection(xmin, ymin, xmax, ymax, score, label)
+                for _, label, score, xmin, ymin, xmax, ymax in outputs[self.output_name][0][0]]
+
 
+class MultipleOutputParser:
+    def __init__(self, all_outputs, bboxes_layer='bboxes', scores_layer='scores', labels_layer='labels'):
+        self.labels_layer = self.find_layer(labels_layer, all_outputs)
+        self.scores_layer = self.find_layer(scores_layer, all_outputs)
+        self.bboxes_layer = self.find_layer(bboxes_layer, all_outputs)
 
-class MultipleOutputPostprocessor:
-    def __init__(self, bboxes_layer='bboxes', scores_layer='scores', labels_layer='labels'):
-        self.bboxes_layer = bboxes_layer
-        self.scores_layer = scores_layer
-        self.labels_layer = labels_layer
+    @staticmethod
+    def find_layer(name, all_outputs):
+        suitable_layers = [layer_name for layer_name in all_outputs if name in layer_name]
+        if not suitable_layers:
+            raise ValueError('Suitable layer for "{}" output is not found'.format(name))
+
+        if len(suitable_layers) > 1:
+            raise ValueError('More than 1 layer matched to "{}" output'.format(name))
+
+        return suitable_layers[0]
+
+    def __call__(self, outputs):
+        bboxes = outputs[self.bboxes_layer][0]
+        scores = outputs[self.scores_layer][0]
+        labels = outputs[self.labels_layer][0]
+        return [Detector.Detection(*bbox, score, label) for label, score, bbox in zip(labels, scores, bboxes)]
+
+
+class OTEParser:
```

Comment:
Why is this called `OTEParser`? Presumably OTE stands for OpenVINO training extensions, but surely not all OTE models have this form of output. I would suggest giving it a more descriptive name.

Reply:
We're on a way to move all (maybe some legacy models will be kept as is, but the majority of models are affected) detection models from OTE (OpenVINO Training Extensions) to a common engine (PyTorch -> Object Detection). Face and person detection models are already there and more to come.

Reply:
Sure, but I have no reason to believe that you will stay with that engine forever (or that the engine will always produce outputs in this format). At the very least, if this format comes from that engine, then the parser should be named after the engine, not OTE.